ggplot plotting API

yafimvo opened the pull request:
## Describe your changes
This is the 1st example of the ggplot API in jupysql.

1. ggplot api added
2. boxplot and histogram added
3. custom style added to histogram
4. stacked bar plot added
5. `facet_wrap` added
6. Example added
7. Test cases added

## How to use it:
The `ggplot` API is structured around the principles of the grammar of graphics, and it comprises three primary components:

`data` - our SQL tables
`aes` - aesthetic properties (x and y values)
`geom` - includes the geometric objects such as `geom_boxplot` and `geom_histogram`

To use it, we first should initialize a `ggplot` instance and reference our data using the `table` parameter, and set `aes` object by specifying the relevant values, `x` and/or `y`.

```python
ggplot('table-name', aes(x='col_from_table', y='another_col_from_table'))
```

The final steps is to define which plot we want to use, `geom_boxplot` or `geom_histogram` and add it to our `ggplot` instance.

```python
ggplot('table-name', aes(x='col_from_table', y='another_col_from_table')) + geom_histogram(bins = 10)
```

We can specify the bar color and its edges with the `fill` and `color` parameters.

```python
ggplot('table-name', aes(x='col_from_table', y='another_col_from_table')) + geom_histogram(bins = 10, fill='red', color='white')
```

## Issue ticket number and link
Closes #148 

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--226.org.readthedocs.build/en/226/

<!-- readthedocs-preview jupysql end -->

yafimvo commented:
@edublancas Please let me know if it is the right direction and what more features/geoms we want to add

edublancas commented on doc/user-guide/ggplot.md:
```diff
@@ -0,0 +1,91 @@
+(
+    ggplot(table="short-trips", with_="short-trips")
+    + aes(x="trip_distance") # noqa
+    + geom_histogram(bins=10, color="#69f0ae", edgecolor="#fff") # noqa
```
let's match the names to the ones in ggplot, for example, in ggplot instead of color, they use `fill`.

edublancas commented on src/sql/ggplot/__init__.py:
```diff
@@ -0,0 +1,13 @@
+from .ggplot import ggplot
```
let's switch for absolute imports, relative imports are discouraged

edublancas reviewed (changes requested):
this looks great overall!

I think we can now implement some basic customizations. Example, for histograms, it's pretty common to generate them by splitting by some categorical variable

![image](https://user-images.githubusercontent.com/989250/224113106-e49cd29d-8347-4eb4-bb57-0361c6c0110f.png)

[source](https://ggplot2.tidyverse.org/reference/geom_histogram.html)

the other similar one is again using a categorical variable but creating a grid:

![image](https://user-images.githubusercontent.com/989250/224113230-12b56fe4-d368-4883-a2ef-9cc9e00e2bb6.png)

so for this we can modify the SQL queries, grab the data and then use matplotlib to creat the plots

note that this examples use binwidth instead of n_bins; we can implement that later, for now, let's stick with n_bins

edublancas reviewed (changes requested):
this is great! I think we can ship it this week!

comments:

Since this is a new API and we want to encourage user feedback: I'm thinking we can put this under a new documentation section. Maybe call it "Experimental"? The objective is to tell the users that this is a new API and is subject to change (hence, they should expect some breaking API changes without the usual notice we give when breaking the API) and we should encourage them to share their feedback on Slack.

Also, there should be a versionadded admonition at the top of the tutorial telling users that this was added on version 0.7

Is there anything here that's not standard in R's ggplot? If so, we should tell the users. I think the only difference is that we have a 'tables' argument, right? Is there anything else (argument names, types, etc)

edublancas commented:
oh, and there is a merge conflict

edublancas commented:
hi @yafimvo there are some merging conflicts since we merged a big PR from @tonykploomber

edublancas commented:
> Maybe plotting section instead of experimental.

but the idea is to communicate that this is an API subject to change

edublancas commented:
> I think it's better to specify it in a note than calling the section/work we done experimental (it might create antagonistic response/make some users flinch instead of trying it)

fair point. then I think we can leave it where it is and just add an admonition at the top of the section

yafimvo commented:
Based on the feedback, I added some changes to the API to make sure it matches the original API of `ggplot2` (color and fill are now part of aes, and cmap and fill are part of geom. the overlap between aes's fill and geom's fill are in the plotline API. geom's fill automatically stacks the bar by it's value) and changed the docs (added the release note and a short note that this is the 1st release and we encourage people to share their feedback with us).

edublancas commented on src/sql/ggplot/__init__.py:
```diff
@@ -0,0 +1,15 @@
+from sql.ggplot.geom import geom_boxplot, geom_histogram
+from sql.ggplot.facet_wrap import facet_wrap
+
+__version__ = "0.6.4dev"
```
```suggestion
```

the only file that should have `__version__` is the root  init file

edublancas commented on src/sql/ggplot/geom/__init__.py:
```diff
@@ -0,0 +1,10 @@
+from sql.ggplot.geom.geom_boxplot import geom_boxplot
+from sql.ggplot.geom.geom_histogram import geom_histogram
+
+__version__ = "0.6.4dev"
```
```suggestion

```

edublancas commented on src/sql/ggplot/geom/geom_boxplot.py:
```diff
@@ -0,0 +1,22 @@
+from sql import plot
+from .geom import geom
```
relative import are discouraged

edublancas commented on src/sql/ggplot/geom/geom_histogram.py:
```diff
@@ -0,0 +1,41 @@
+from sql import plot
+from .geom import geom
```
relative import

edublancas commented on src/sql/plot.py:
```diff
@@ -7,18 +7,20 @@
 
 try:
     import matplotlib.pyplot as plt
+    from matplotlib.colors import Normalize
```
since matplotlib is optional, let's add the requires decorator to the `ggplot` object so it raises an error asking to install matplotlib: https://ploomber-core.readthedocs.io/en/latest/dependencies.html

edublancas commented on src/sql/plot.py:
```diff
@@ -352,37 +474,107 @@ def histogram(payload, table, column, bins, with_=None, conn=None):
+    # TODO: We can use the duckdb pivot feature which will be released
+    # in 0.8 so we won't have to use dataframe
+    # https://github.com/duckdb/duckdb/pull/6387
+    import pandas as pd
```
since matplotlib already uses numpy as dependency, maybe we can use that? https://numpy.org/doc/stable/reference/generated/numpy.transpose.html

if we need to keep it, then let's add pandas to the requires decorator (as mentioned in my previous comment). also, let's move this import to the top and wrap it in a try except statement.

edublancas replied:
and let's also add an admonition in the tutorial saying that ggplot requires matplotlib (and pandas if we decide to keep it)

yafimvo commented on pyproject.toml:
```diff
@@ -16,5 +16,7 @@ flake8 = [
-    "--ignore=F821",
+    # W503, W504 ignore line break after/before 
+    # binary operator since they are conflicting 
+    "--ignore=F821, W503, W504",
```
when I use 1 it says that I need to use the other one

edublancas commented:
yeeeeesss 🔥🔥
